src/intrinsics/ecma262/ObjectPrototype.js

```diff
@@ -10,24 +10,47 @@
 /* @flow */
 
 import type { Realm } from "../../realm.js";
-import { NativeFunctionValue, ObjectValue, BooleanValue, NullValue } from "../../values/index.js";
+import { AbstractValue, NativeFunctionValue, ObjectValue, BooleanValue, NullValue } from "../../values/index.js";
 import { SameValuePartial, RequireObjectCoercible } from "../../methods/abstract.js";
 import { HasOwnProperty, HasSomeCompatibleType } from "../../methods/has.js";
 import { Invoke } from "../../methods/call.js";
 import { Properties, To } from "../../singletons.js";
+import { FatalError } from "../../errors.js";
+import type { BabelNodeExpression } from "babel-types";
+import * as t from "babel-types";
 import invariant from "../../invariant.js";
 
 export default function(realm: Realm, obj: ObjectValue): void {
   // ECMA262 19.1.3.2
-  obj.defineNativeMethod("hasOwnProperty", 1, (context, [V]) => {
-    // 1. Let P be ? ToPropertyKey(V).
-    let P = To.ToPropertyKey(realm, V.throwIfNotConcrete());
-
-    // 2. Let O be ? ToObject(this value).
-    let O = To.ToObjectPartial(realm, context);
-
-    // 3. Return ? HasOwnProperty(O, P).
-    return new BooleanValue(realm, HasOwnProperty(realm, O, P));
+  const ObjectPrototypeHasOwnPrototype = obj.defineNativeMethod("hasOwnProperty", 1, (context, [V]) => {
+    try {
+      // 1. Let P be ? ToPropertyKey(V).
+      let P = To.ToPropertyKey(realm, V.throwIfNotConcrete());
+
+      // 2. Let O be ? ToObject(this value).
+      let O = To.ToObjectPartial(realm, context);
+
+      // 3. Return ? HasOwnProperty(O, P).
+      return new BooleanValue(realm, HasOwnProperty(realm, O, P));
+    } catch (x) {
+      if (realm.isInPureScope() && x instanceof FatalError) {
+        // If we're in pure scope we can try to recover from any fatals by
+        // leaving the call in place which we do by default, but we don't
+        // have to havoc the state of any arguments since this function is pure.
+        // This also lets us define the return type properly.
+        return realm.evaluateWithPossibleThrowCompletion(() =>
+          AbstractValue.createTemporalFromBuildFunction(
+            realm,
+            BooleanValue,
+            [ObjectPrototypeHasOwnPrototype, context, V],
+            ([methodNode, objectNode, nameNode]: Array<BabelNodeExpression>) => {
+              return t.callExpression(t.memberExpression(methodNode, t.identifier("call")), [objectNode, nameNode]);
```

Comment:
I suspect that t.callExpression(methodNode, [objectNode, nameNode]) will work if you pass in ObjectPrototypeHasOwnPrototype.$Call in line 45. That would be better code to generate.

Reply:
The `.call` method is not bound to the `hasOwnProperty` function so that doesn't work because the `call` function needs `this` to be the `ObjectPrototypeHasOwnPrototype` function.

Reply:
To say it another way, this is not valid:

```js
var hasOwnProperty = Object.prototype.hasOwnProperty.call;
hasOwnProperty(obj, key);
```